contracts/btc-finality/src/finality.rs

```diff
@@ -488,6 +488,14 @@ pub fn handle_finality_signature(
     // Store the block height this finality provider has signed
     FP_BLOCK_SIGNER.save(deps.storage, &fp_btc_pk_hex, &height)?;
 
+    // Accumulate voting weight for this FP for reward distribution
+    let power_table = get_power_table_at_height(deps.storage, height)?;
+    if let Some(voting_power) = power_table.get(&fp_btc_pk_hex) {
+        ACCUMULATED_VOTING_WEIGHTS.update(deps.storage, &fp_btc_pk_hex, |existing| {
+            Ok::<u128, ContractError>(existing.unwrap_or(0) + (*voting_power as u128))
+        })?;
+    }
```

Comment:
Getting the entire voting power table is not efficient, with DB read complexity of `O(n)` where `n` is the number of active FPs.

Alternatively, we can revise L399-400 in a way that

- we get voting power of this FP
- If voting power is 0, return error
- use this voting power at L494

Reply:
Good point, we already load voting power table to check voting power in Line 400, we should not load voting power table again